devtools/src/arcs-hot-reload.js

```diff
@@ -0,0 +1,149 @@
+
+class HotReload extends MessengerMixin(PolymerElement) {
+  static get template() {
+    return html`
+    <style include="shared-styles">
+      :host {
+        display: flex;
+        flex-direction: column;
+      }
+      .switch {
+        position: relative;
+        display: inline-block;
+        width: 30px;
+        height: 17px;
+      }
+      .slider {
+        position: absolute;
+        top: 0;
+        left: 0;
+        right: 0;
+        bottom: 0;
+        background-color: #ccc;
+        border-radius: 34px;
+      }
+      .slider:before {
+        position: absolute;
+        content: "";
+        height: 13px;
+        width: 13px;
+        left: 2px;
+        bottom: 2px;
+        background-color: white;
+        -webkit-transition: .4s;
+        border-radius: 50%;
+      }
+      input:checked + .slider {
+        background-color: #AFF196;
+      }
+      input:checked + .slider:before {
+        -webkit-transform: translateX(13px);
+      }
+      .tooltip:hover .tooltiptext {
+        visibility: visible;
+      }
+      .tooltip .tooltiptext {
+        visibility: hidden;
+        width: 120px;
+        text-align: center;
+        color: #838783;
+        font-size: 10px;
+        position: absolute;
+        top: 8px;
+        left: 45%;
+      }
+      .tooltip {
+        position: relative;
+        display: inline-block;
+      }
+    </style>
+    <div class="tooltip">
+      <label class="switch">
+        <input type="checkbox" on-click="_loadServer" id="toggle">
+        <span class="slider round"></span>
+      </label>
+      <span class="tooltiptext">
+        Hot Code Reload
+      </span>
+    </div>`;
+  }
+
+  static get is() { return 'arcs-hot-reload'; }
+
+  constructor() {
+    super();
+    this.hotReloadWS = null;
+    this.hotReloadServer = null;
+    this.particleSources = new Set();
+    this.arcIdsToKeepFresh = [];
+  }
+
+  onMessage(msg) {
+    switch (msg.messageType) {
+      case 'register-arc':
+        this.arcIdsToKeepFresh.push(msg.messageBody);
+        break;
+      case 'register-particle-sources': {
+        const before = this.particleSources.size;
+        for (const particleSource of msg.messageBody) {
+          this.particleSources.add(particleSource);
+        }
+        if (this.particleSources.size > before && this.hotReloadWS != null) {
+          const files = JSON.stringify([...this.particleSources]);
+          this.hotReloadWS.send(files);
+        } 
+        break;
+      }
+      case 'particle-reload-result':
+        break;
+    }
+  }
+
+  _loadServer() {
+    if (this.$.toggle.checked) {
+      this.$.toggle.checked = false;
+
+      this.hotReloadWS = new WebSocket('ws://localhost:8888');
+      this.hotReloadWS.onopen = e => {
+        this.hotReloadWS.send(JSON.stringify([...this.particleSources]));
+
+        this.hotReloadWS.onmessage = msg => {
+          // send message to devtool inspector through connector
```

Comment:
Maybe instead say what this is doing:

"Notify each arc of the particle source code file which got updated"

We could debate whether instead we should send a single message and go over the arcs on the runtime side - but it doesn't really matter. I think sometime soon we will provide more powerful control of what gets reloaded and what doesn't, then we can refactor it to how we want it.

Reply:
Changed! I also don't really mind if we want to keep track all the arcs and then iterate over them in the runtime! :)